cf_remote/utils.py

```diff
@@ -226,15 +228,64 @@ def print_progress_dot(*args):
     sys.stdout.flush()  # STDOUT is line-buffered
 
 
+# atomic copy, see lock-free whack-a-mole algorithm
 def copy_file(input_path, output_path):
-    filename = os.path.basename(input_path)
-    output_dir = os.path.dirname(output_path)
-
-    tmp_filename = ".{}.tmp".format(filename)
-    tmp_output_path = os.path.join(output_dir, tmp_filename)
-
-    shutil.copyfile(input_path, tmp_output_path)
-    os.rename(tmp_output_path, output_path)
+    if os.path.exists(output_path):
+        return
+
+    output_filename = os.path.basename(output_path)
+    if not output_filename:
+        output_filename = os.path.basename(input_path)
+
+    output_dirname = os.path.dirname(output_path)
+    my_uuid = uuid.uuid4()
+
+    tmp_path = os.path.join(
+        output_dirname, "{}-{}.tmp".format(output_filename, my_uuid)
+    )
+    shutil.copyfile(input_path, tmp_path)
```

Comment:
The temporary file should be opened with the `O_EXCL` flag _(see `man 2 open`)_, to make sure it did not already exist process. Otherwise, two processes could coincidentally generate a temporary file with the same filename.

The `mkstemp()` function does this for you. So I would recommend using it.